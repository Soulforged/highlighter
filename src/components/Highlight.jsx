//@flow
import React from "react";

type Props = {
  x: number,
  y: number,
  width: number,
  color: string
}

export default ({ x, y, color, width, id }: Props) => (
  <mark
    data-testid={id}
    style={{
      position: "absolute",
      left: x + 1,
      top: y - 6,
      backgroundColor: color,
      color: "transparent",
      width: width,
      height: "1.1em"
    }}
  />
);
