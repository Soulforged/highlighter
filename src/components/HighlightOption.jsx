//@flow
import React from "react";

type Props = {
  color: string,
  onSelected: (string) => void,
  selected: boolean
};

export default ({ color, selected, onSelected }: Props) => (
  <button
    style={{
      width: 50,
      height: 32,
      margin: 5,
      backgroundColor: color,
      border: `3px solid ${selected ? "blue" : "black" }`
    }}
    onClick={onSelected}
  />
);
