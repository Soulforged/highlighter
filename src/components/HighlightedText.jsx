//@flow
import React from "react";

type Props = {
  width: number,
  color: string,
  text: string
}

export default ({ color, width, text }: Props) => (
  <p
    style={{
      backgroundColor: color,
      width: width
    }}
  >
    {text}
  </p>
);
