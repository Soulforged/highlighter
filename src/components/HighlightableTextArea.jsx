//@flow
import React, { useState } from "react";
import Highlight from "./Highlight";

function getTextBoundingRect(input, selectionStart, selectionEnd, debug) {
    // If available (thus IE), use the createTextRange method
    if (typeof input.createTextRange == "function") {
        var range = input.createTextRange();
        range.collapse(true);
        range.moveStart('character', selectionStart);
        range.moveEnd('character', selectionEnd - selectionStart);
        return range.getBoundingClientRect();
    }
    // createTextRange is not supported, create a fake text range
    let offset = getInputOffset();
    let topPos = offset.top;
    let leftPos = offset.left;
    let width = getInputCSS('width', true);
    let height = getInputCSS('height', true);

    // Styles to simulate a node in an input field
    let cssDefaultStyles = "white-space:pre;padding:0;margin:0;";
    let listOfModifiers = ['direction', 'font-family', 'font-size',
          'font-size-adjust', 'font-variant', 'font-weight', 'font-style',
          'letter-spacing', 'line-height', 'text-align', 'text-indent',
          'text-transform', 'word-wrap', 'word-spacing'];

    topPos += getInputCSS('padding-top', true);
    topPos += getInputCSS('border-top-width', true);
    leftPos += getInputCSS('padding-left', true);
    leftPos += getInputCSS('border-left-width', true);
    leftPos += 1; //Seems to be necessary

    for (var i=0; i < listOfModifiers.length; i++) {
        var property = listOfModifiers[i];
        cssDefaultStyles += property + ':' + getInputCSS(property) +';';
    }
    // End of CSS variable checks

    const text = input.value;
    const textLen = text.length;
    const fakeClone = document.createElement("div");
    if(selectionStart > 0) {
      appendPart(0, selectionStart);
    }
    var fakeRange = appendPart(selectionStart, selectionEnd);
    if(textLen > selectionEnd) {
      appendPart(selectionEnd, textLen);
    }

    // Styles to inherit the font styles of the element
    fakeClone.style.cssText = cssDefaultStyles;

    // Styles to position the text node at the desired position
    fakeClone.style.position = "absolute";
    fakeClone.style.top = topPos + "px";
    fakeClone.style.left = leftPos + "px";
    fakeClone.style.width = width + "px";
    fakeClone.style.height = height + "px";
    fakeClone.style.color = "transparent";
    fakeClone.style["z-index"] = -4;
    document.getElementById("root").appendChild(fakeClone);
    var returnValue = fakeRange.getBoundingClientRect(); //Get rect

    if (!debug) {
      fakeClone.parentNode.removeChild(fakeClone);//Remove temp
    }
    return returnValue;

    // Local functions for readability of the previous code
    function appendPart(start, end){
        var span = document.createElement("span");
        span.style.cssText = cssDefaultStyles; //Force styles to prevent unexpected results
        span.textContent = text.substring(start, end);
        fakeClone.appendChild(span);
        return span;
    }
    // Computing offset position
    function getInputOffset(){
        var body = document.body,
            win = document.defaultView,
            docElem = document.documentElement,
            box = document.createElement('div');
        box.style.paddingLeft = box.style.width = "1px";
        body.appendChild(box);
        var isBoxModel = box.offsetWidth == 2;
        body.removeChild(box);
        box = input.getBoundingClientRect();
        var clientTop  = docElem.clientTop  || body.clientTop  || 0,
            clientLeft = docElem.clientLeft || body.clientLeft || 0,
            scrollTop  = win.pageYOffset || isBoxModel && docElem.scrollTop  || body.scrollTop,
            scrollLeft = win.pageXOffset || isBoxModel && docElem.scrollLeft || body.scrollLeft;
        return {
            top : box.top  + scrollTop  - clientTop,
            left: box.left + scrollLeft - clientLeft};
    }
    function getInputCSS(prop, isnumber){
        var val = document.defaultView.getComputedStyle(input, null).getPropertyValue(prop);
        return isnumber ? parseFloat(val) : val;
    }
}

const renderHighlights = highlights => (
  highlights.map(({ x, y, width, height, color, id }) => (
    <Highlight
      key={id}
      width={width}
      height={height}
      x={x}
      y={y}
      color={color}
      id={id}
    />
  ))
);

export default ({
  highlights = [],
  addHighlight,
  highlightColor,
  width = 500,
  height = 100
}) => {
  const [selectionStartX,setSelectionStartX] = useState(0);
  const calculateAndAddHighlight = ({ target, nativeEvent }) => {
    const { selectionStart, selectionEnd } = target;
    // getTextBoundingRect(target, selectionStart, selectionEnd, true);
    // const { offsetX, offsetY } = nativeEvent;
    const highlight = target.value.substring(selectionStart, selectionEnd);
    // console.log(selectionStartX, selectionEnd, highlight, offsetX);
    // if (highlight && offsetX !== undefined && offsetY !== undefined){
    if (highlight){
      const rect = getTextBoundingRect(target, selectionStart, selectionEnd, true)
      // const width = offsetX - selectionStartX;
      const id = `${highlightColor}_${rect.x}_${rect.y}`;
      addHighlight({
        text: highlight,
        x: rect.x,
        y: rect.y,
        width: rect.width,
        color: highlightColor,
        id
      });
    }
  };
  return (
    <div style={{ position: "relative" }}>
      <div style={{
        backgroundColor: "white",
        overflow: "auto",
        position: "absolute",
        height,
        width,
        top: 0,
        left: 0,
        zIndex: -1,
        margin: 0,
        borderRadius: 0,
        padding: 3
      }}>
        {renderHighlights(highlights)}
      </div>
      <textarea
        style={{
          width,
          height,
          backgroundColor: "transparent",
          margin: 0,
          borderRadius: 0
        }}
        data-testid="highlightableTxtArea"
        onMouseDown={({ nativeEvent: { offsetX } }) => setSelectionStartX(offsetX)}
        onSelect={calculateAndAddHighlight}
      />
    </div>
  );
}
