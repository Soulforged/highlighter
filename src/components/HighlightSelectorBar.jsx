//@flow
import React from "react";
import HighlightOption from "./HighlightOption";

export default ({ highlighterColors, setHighlightColor, highlightColor }) => (
  <div style={{ flexDirection: "row", padding: 10 }}>
    {highlighterColors.map(color => (
      <HighlightOption
        key={color}
        onSelected={() => setHighlightColor(color)}
        color={color}
        selected={highlightColor === color}
      />
    ))}
  </div>
);
