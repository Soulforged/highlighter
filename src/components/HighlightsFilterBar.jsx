//@flow
import React from "react";
import HighlightOption from "./HighlightOption";

export default ({
  highlighterColors,
  selectedHighlightOptions,
  removeSelectedHighlightOption,
  addSelectedHighlightOption
}) => (
  <div style={{ flexDirection: "row", padding: 10 }}>
    {highlighterColors.map(color => (
      <HighlightOption
        key={color}
        onSelected={() => {
          if (selectedHighlightOptions.indexOf(color) !== -1) {
            removeSelectedHighlightOption(color);
          } else {
            addSelectedHighlightOption(color);
          }
        }}
        color={color}
        selected={selectedHighlightOptions.indexOf(color) !== -1}
      />
    ))}
  </div>
);
