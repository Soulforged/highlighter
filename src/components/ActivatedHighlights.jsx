//@flow
import React from "react";
import HighlightedText from "./HighlightedText";

const renderActivatedHighlights = filteredHighlights => (
  filteredHighlights.map(({ text, width, color, id }) =>
    <HighlightedText key={id} text={text} width={width} color={color} />
  )
);

export default ({ filteredHighlights }) => (
  <div style={{
    overflow: "auto",
    width: 500,
    height: 100,
    padding: 3,
    border: "1px solid black"
  }}>
    {renderActivatedHighlights(filteredHighlights)}
  </div>
);
