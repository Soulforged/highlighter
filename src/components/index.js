import HighlightableTextArea from "./HighlightableTextArea";
import HighlightSelectorBar from "./HighlightSelectorBar";
import ActivatedHighlights from "./ActivatedHighlights";
import HighlightsFilterBar from "./HighlightsFilterBar";

export {
  HighlightableTextArea,
  HighlightSelectorBar,
  ActivatedHighlights,
  HighlightsFilterBar
};
