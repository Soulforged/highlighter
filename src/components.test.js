import React from 'react';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
  waitForDomChange,
  wait
} from "@testing-library/react";
import { HighlightableTextArea } from './components';

describe("component tests", () => {
  const highlights = [{
    text: "some text",
    x: 0,
    y: 1,
    width: 200,
    color: "red",
    id: "red_0_1",
    key: "red_0_1"
  }];

  it("renders highlights when text gets selected in the a HighlightableTextArea", () => {
    const { container, getByTestId } = render((
      <HighlightableTextArea highlights={highlights} />
    ));
    const mark = getByTestId("red_0_1");
    expect(mark).not.toBe(null);
    expect(container).toMatchSnapshot();
  });

  it("allows to add new highlights by selecting text over HighlightableTextArea", () => {
    const addHighlightMock = jest.fn();
    const { getByTestId } = render((
      <HighlightableTextArea
        addHighlight={addHighlightMock}
        highlightColor="red"
      />
    ));
    const textArea = getByTestId("highlightableTxtArea");
    fireEvent.change(textArea, { target: { value: "Lorem ipsum dolor" } });
    fireEvent.mouseDown(textArea, { nativeEvent: { offsetX: 2 } });
    fireEvent.select(textArea, {
      target: { selectionStart: 3, selectionEnd: 8 },
      nativeEvent: { offsetX: 3, offsetY: 2 }
    })
    expect(addHighlightMock).toHaveBeenCalledWith({
      text: "em ipsum ",
      x: 3,
      y: 8,
      width: 200,
      color: "red",
      id: "red_0_1"
    });
  });
});
