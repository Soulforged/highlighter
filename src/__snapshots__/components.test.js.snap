// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component tests renders highlights when text gets selected in the a HighlightableTextArea 1`] = `
<div>
  <div
    style="position: relative;"
  >
    <div
      style="background-color: white; overflow: auto; position: absolute; height: 100px; width: 500px; top: 0px; left: 0px; z-index: -1; margin: 0px; border-radius: 0; padding: 3px;"
    >
      <mark
        data-testid="red_0_1"
        style="position: absolute; left: 1px; top: -5px; background-color: red; width: 200px; height: 1.1em;"
      />
    </div>
    <textarea
      data-testid="highlightableTxtArea"
      style="width: 500px; height: 100px; background-color: transparent; margin: 0px; border-radius: 0;"
    />
  </div>
</div>
`;
